#171380755 Email and in-app notifications

Muhire-Josue opened the pull request:
#### What does this PR do?

- It provides both in-app and Email notifications upon user create trip request

#### Description of Task to be completed?

- When a user creates a trip request they get an email notification about their trip request.

- A user can also disable email notifications so that they don't receive email notifications anymore.

- A manager is able to get the in-app notifications.

#### How should this be manually tested?

- `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git`
- `cd stackup2-barefoot-backend`
- `git fetch origin ft-email-inapp-notifications-171380755`
-`git checkout ft-email-inapp-notifications-171380755`
- `npm install`
- Use `.env.example` as reference for environment variables
- In case you don't have a functional database for this project, run `npm run create-db`
- `npm run migrations`
- `npm run dev`
- At this point, the server must be running and the application should be connected to the database and all the tables in the database are created.
- Signup using this URL `localhost:3000/api/auth/signup` use this sample object 
```
{
	"firstName": "YOUR_FIRST_NAME",
	"lastName":"YOUR_LAST_NAME",
	"address":"YOUR_ADDRESS",
	"username":"YOUR_USERNAME",
	"password":"YOUR_PASSWORD",
	"email":"YOUR_EMAIL",
	"gender":"YOUR_GENDER"
}
```
- Login in order to verify your account using this URL `localhost:3000/api/auth/login` use this sample object

```
"email":"YOUR_EMAIL",
"password":"YOUR_PASSWORD"

```
- create a trip using this URL `localhost:3000/api/trips` use this object sample 
```
{
    "travelFrom": "kigali",
    "travelTo": "Butare",
    "travelDate": "2020-05-14",
    "travelReason": "business meeting",
    "travelType": "one-way",
    "accommodation": true
}

```

- Use this URL to disable email notifications, PATCH: `http://localhost:3000/api/auth/disable?mode=email`

- Use this URL to disable in-appnotifications, PATCH: `http://localhost:3000/api/auth/disableOrEnable?mode=inapp`

-  Use this URL to get in-app notifications `localhost:3000/api/trips/inapp/notification/`
- Use this URL to get unread in-app notifications `localhost:3000/api/trips/inapp/notification?status=unread`

#### Any background context you want to provide?

- When you signup please provide a real email address that you can access because notifications will be sent that email address that you provided during signup.

- In order to get in-app notifications you have to be a manager.

#### What are the relevant pivotal tracker stories?

- [171380755](https://www.pivotaltracker.com/story/show/171380755)

#### Screenshots (if appropriate)
![inapp3](https://user-images.githubusercontent.com/37122177/81540666-4451fc80-9372-11ea-8f03-30e61648c7a7.PNG)
![inapp2](https://user-images.githubusercontent.com/37122177/81540676-46b45680-9372-11ea-8eb6-7ddd55454e4e.PNG)
![inapp1](https://user-images.githubusercontent.com/37122177/81540677-47e58380-9372-11ea-9651-008909963853.PNG)
![requesterEmail](https://user-images.githubusercontent.com/37122177/81540679-487e1a00-9372-11ea-98ca-680fd060d83c.PNG)
![managerEmail](https://user-images.githubusercontent.com/37122177/81540683-4916b080-9372-11ea-828a-b0efdb2cb372.PNG)
![nonManager](https://user-images.githubusercontent.com/37122177/81540687-49af4700-9372-11ea-86b6-873421f5109c.PNG)
![managerDontExist](https://user-images.githubusercontent.com/37122177/81540691-4ae07400-9372-11ea-9d5e-eacdcba9dd5a.PNG)

Ugizwenayo-Divine reviewed (commented):
Good job Josue
Would you please solve those 3 issues

Muhire-Josue committed:
feature(notification): add email and inapp notification

- add email notification upon trip request features
- add filter inapp notifications
- add inapp notification for managers
- add view details of a trip request from the email
- add opt in/out of notifications
- add tests

[Finishes #171380755]

Ugizwenayo-Divine reviewed (commented):
Would you help me with this issue
I am getting this error while creating a trip request. 
![Josue](https://user-images.githubusercontent.com/52660542/81044640-6d452f80-8e50-11ea-9cd8-9096c40e663a.PNG)

Ugizwenayo-Divine reviewed (commented):
I would suggest to put enough description on your PR:
- what's the problem here
![Josue](https://user-images.githubusercontent.com/52660542/81081123-38a39900-8e8d-11ea-9dd6-8fffb3b708ea.PNG)

Ugizwenayo-Divine reviewed (commented):
Please, would you solve CI failing issue and also the code climate issue

the22mastermind commented on README.md:
```diff
@@ -114,4 +114,7 @@ Refer to the [sequelize](https://sequelize.org/v5/) and [sequelize-cli](https://
 | /api/profile/password | PATCH | 200 OK | Makes a PATCH request to help password changes, whenever a user needs to change his/her password directly from his/her profile|
 | /api/auth/verify?token=eyWWWSDHFGC | GET | 200 OK | Makes a GET request when a user clicks on the link that is sent to him/her via Email to verify his/her email |
+| /api/trips/inapp/notification | POST | 200 OK | It shows the inApp notifications to manager and requester |
+| api/auth/disable?mode=inapp | POST | 200 OK | It enables/disables inApp notifications |
```
Please change the request method to GET.

the22mastermind commented on README.md:
```diff
@@ -114,4 +114,7 @@ Refer to the [sequelize](https://sequelize.org/v5/) and [sequelize-cli](https://
 | /api/auth/verify?token=eyWWWSDHFGC | GET | 200 OK | Makes a GET request when a user clicks on the link that is sent to him/her via Email to verify his/her email |
+| /api/trips/inapp/notification | POST | 200 OK | It shows the inApp notifications to manager and requester |
+| api/auth/disable?mode=inapp | POST | 200 OK | It enables/disables inApp notifications |
+| api/auth/disable?mode=email | POST | 200 OK | It enables/disables email notifications |
```
Same here

the22mastermind commented on src/controllers/authentication.controller.js:
```diff
@@ -24,7 +26,7 @@ const {
   findUserEmailIfExist,
   updateUserPassword,
-  updateIsVerified
+  updateIsVerifiedOrDisableNotification
```
Could you rename this function to maybe `updateIsVerifiedOrUpdateNotification` since the part of notifications deals with both disable and enable?

the22mastermind commented on src/controllers/comment.controller.js:
```diff
@@ -40,10 +45,12 @@ export default class CommentController {
         const { requestId, comment } = req.body;
         const commentData = { id: commentId, requestId, comment };
         const updatedCommentToReturn = await updateComment(commentData);
+        await createInAppNotification(currentUser, requestId, 4);
```
Could you use the variables you created in `notificationTypes.utils` instead of `4`?

the22mastermind commented on src/controllers/notification.controller.js:
```diff
@@ -0,0 +1,52 @@
+    if (!canViewNotification) {
+      return errorResponse(res, unAuthorized, notificationsDisabled);
+    }
+    if (role === 'manager') {
```
Instead of `manager` use the variable in userRoles.utils.

the22mastermind commented on src/services/authentication.service.js:
```diff
@@ -106,17 +113,35 @@ export default class UserService {
+   * @param {boolean} status
+   * @returns {object} returns an updated user
+   */
+  static updateIsVerifiedOrDisableNotification = async (value, status) => {
```
Could you change the name of this method and update it's JSDoc to reflect what it does exactly?

the22mastermind commented on src/services/request.service.js:
```diff
@@ -198,4 +198,45 @@ export default class RequestService {
+   * @param {string} ids
+   * @returns {Object} an array of requests based on the array of ids passed through params
+    */
+  // static async getTripRequestsFromRequesterId(ids) {
```
Remove this commented code.

the22mastermind commented on src/utils/disableNotification.util.js:
```diff
@@ -0,0 +1,19 @@
+
+const { updateIsVerifiedOrDisableNotification, getUserById } = userService;
+const { updateInAppNotification } = requestService;
+const disableNotification = async (mode, id) => {
```
Could you add empty lines before this function and after it, to separate it from the destructured statements above and the export statement below?

the22mastermind commented on src/utils/getInappNotification.js:
```diff
@@ -0,0 +1,46 @@
+ * @description this function returns in-app notifications for manager
+ * details depending on the value passed
+ */
+const getInAppNotificationForManager = async (id, notificationStatus) => {
```
Please remove the commented code in this function.

the22mastermind commented on tests/controllers/notifications.test.js:
```diff
@@ -0,0 +1,31 @@
+const { created } = statusCodes;
+let userToken;
+
+describe('Notifications Tests', () => {
```
Could you delete this file since it's not being used?

the22mastermind reviewed (commented):
@Muhire-Josue Great job!
Please check out the comments I left.
Additionally, the endpoint for enabling and disabling notifications uses GET. Don't you think it should use PATCH? Also check this endpoint in your PR description, your README and in routes to make sure they are matching.

Finally, I think this message should be changed to maybe `Here are all your unread notifications` since it's them we are retrieving not trip requests.
![Screenshot from 2020-05-10 20-48-40](https://user-images.githubusercontent.com/45935765/81510352-4f287500-9311-11ea-8f2c-2bea75519c79.png)

descholar-ceo reviewed (dismissed):
This is great @Muhire-Josue, you have done a tremendous job, special congrats. However, I have noticed something we talked about on standup, you are still using that old way of writing repeatedly methods of interacting with the DB. Could you move those methods in CrudRepository object?

descholar-ceo reviewed (changes requested):
1. It is difficult to find the endpoints you worked on. I guess you documented them in swagger and readme files but you should make your PR description holds all necessary information. So could you add the endpoints in your PR description? Refer to this documentation [Andela best practices](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices), take a look on the PR section

2. In your PR's `how should this be manually tested` section, the steps you provided don't lead to your branch testing, because after cloning the repo, (wht you clone is the develop branch) not yours. So, all of the steps you provided lead to testing of develop branch. Could you take care of that?
![Screenshot from 2020-05-13 11-34-12](https://user-images.githubusercontent.com/56886064/81796440-bad73180-950d-11ea-9565-0241c47ec6d3.png)

descholar-ceo reviewed (changes requested):
@Muhire-Josue congrats that you worked hard to realise this task, however, I have few concerns:

A. It looks like you did these three endpoints : 
`GET      /api/auth/disable?mode=inapp`
`GET     /api/trips/inapp/notification/`
`GET    /api/trips/inapp/notification?status=unread`

===> Since there is an endpoint of disabling notifications, then we need another one for enabling them back. Could you work on it too?

B. I think whenever a requester creates a request, the barefoot nomad was supposed to send him email notification, I created one but I did not get any email, are you sure that it works? If so, what did I miss? 

### INFO: 

1. When I signed up I used my live email `emmanuellamugi@gmail.com` this is a requester who requested the trip

2. The manager of that requester email is this one `emmamugira@gmail.com` which is my email too and I use it, (means it is valid), but this manager did not get the email notification

C. I read your endpoint documentation from README.md file and I found this: 
![Screenshot from 2020-05-13 12-30-39](https://user-images.githubusercontent.com/56886064/81802007-9717e980-9515-11ea-8c53-ec374bb28dca.png)
Are you sure all of those endpoints are POST? If so, you should update you PR description, it does not match that documentation.


D. Finally, what is the use of this table? 
![Screenshot from 2020-05-13 13-18-35](https://user-images.githubusercontent.com/56886064/81806146-5c657f80-951c-11ea-994c-50fb0ad1a5d3.png)



===> Except those comments above, other things are cool, congratulation again for making it this far!

descholar-ceo reviewed (commented):
Could you solve the conflicts this branch has with the base branch?

the22mastermind commented:
@Muhire-Josue let me know when you finish resolving those conflicts I review your PR.

Ugizwenayo-Divine reviewed (commented):
great job Josue
but would you help me with those route:
- PATCH `http://localhost:3000/api/trips/inapp/notification?status=unread` I'm getting an error `The requested resource was not found`
-GET  `http://localhost:3000/api/auth/disable?mode=email` I'm getting the same error

Ugizwenayo-Divine reviewed (commented):
Well done!!!
- Please update your PR description
- rebase also
LGTM

Ugizwenayo-Divine reviewed (approved):
Good job

kipropbrian commented on src/services/request.service.js:
```diff
@@ -89,6 +90,126 @@ const getModel = (field) => {
   }
-}
 
+  /**
+*@description update open trip request details in database
+* @param {Object} tripRequestData trip request new data
+* @param {Integer} requestId trip request id
+* @returns {Object} updated trip request details
+ */
+  //  updateTripRequest = (tripRequestData, requestId) => {
+  //   if (tripRequestData.travelType === 'one-way') {
+  //     tripRequestData.returnDate = null;
+  //   }
+  //   const {
+  //     travelTo,
+  //     travelFrom,
+  //     travelReason,
+  //     travelType,
+  //     travelDate,
+  //     returnDate,
+  //     accommodation
+  //   } = tripRequestData;
+  //   return request.update(
+  //     {
+  //       travelTo,
+  //       travelFrom,
+  //       travelReason,
+  //       travelType,
+  //       travelDate,
+  //       returnDate,
+  //       accommodation
+  //     },
+  //     { where: { id: requestId } }
```
remember to remove comments

kipropbrian commented on tests/profile/profile.test.js:
```diff
@@ -321,6 +321,7 @@ describe('Update Profile Tests - Line Manager', () => {
         password: requester3Account.password,
       })
       .end((err, res) => {
+        console.log('==>', res.body);
```
No need for this

kipropbrian reviewed (commented):
why was package.lock deleted

kipropbrian reviewed (approved):
Why are you deleting the package-lock?
